narwhals/_duckdb/dataframe.py

```diff
@@ -1,148 +1,172 @@
-    if duckdb_dtype == "BOOLEAN":
-        return dtypes.Boolean()
-    if duckdb_dtype == "INTERVAL":
-        return dtypes.Duration()
-    if duckdb_dtype.startswith("STRUCT"):
-        matchstruc_ = re.findall(r"(\w+)\s+(\w+)", duckdb_dtype)
-        return dtypes.Struct(
-            [
-                dtypes.Field(
-                    matchstruc_[i][0],
-                    native_to_narwhals_dtype(matchstruc_[i][1], version),
-                )
-                for i in range(len(matchstruc_))
-            ]
-        )
-    if match_ := re.match(r"(.*)\[\]$", duckdb_dtype):
-        return dtypes.List(native_to_narwhals_dtype(match_.group(1), version))
-    if match_ := re.match(r"(\w+)\[(\d+)\]", duckdb_dtype):
-        return dtypes.Array(
-            native_to_narwhals_dtype(match_.group(1), version),
-            int(match_.group(2)),
-        )
-    if duckdb_dtype.startswith("DECIMAL("):
-        return dtypes.Decimal()
-    return dtypes.Unknown()  # pragma: no cover
 
 
-class DuckDBInterchangeFrame:
+class DuckDBLazyFrame:
     _implementation = Implementation.DUCKDB
 
     def __init__(
-        self, df: Any, *, backend_version: tuple[int, ...], version: Version
+        self,
+        df: duckdb.DuckDBPyRelation,
+        *,
+        backend_version: tuple[int, ...],
+        version: Version,
     ) -> None:
-        self._native_frame = df
+        self._native_frame: duckdb.DuckDBPyRelation = df
         self._version = version
         self._backend_version = backend_version
         validate_backend_version(self._implementation, self._backend_version)
 
-    def __narwhals_dataframe__(self) -> Any:
+    def __narwhals_dataframe__(self) -> Any:  # pragma: no cover
+        # Keep around for backcompat.
+        if self._version is not Version.V1:
+            msg = "__narwhals_dataframe__ is not implemented for DuckDBLazyFrame"
+            raise AttributeError(msg)
+        return self
+
+    def __narwhals_lazyframe__(self) -> Any:
         return self
 
     def __native_namespace__(self: Self) -> ModuleType:
         return get_duckdb()  # type: ignore[no-any-return]
 
+    def __narwhals_namespace__(self) -> DuckDBNamespace:
+        from narwhals._duckdb.namespace import DuckDBNamespace
+
+        return DuckDBNamespace(
+            backend_version=self._backend_version, version=self._version
+        )
+
     def __getitem__(self, item: str) -> DuckDBInterchangeSeries:
         from narwhals._duckdb.series import DuckDBInterchangeSeries
 
         return DuckDBInterchangeSeries(
             self._native_frame.select(item), version=self._version
         )
 
+    def collect(self) -> Any:
+        import pyarrow as pa  # ignore-banned-import()
+
+        from narwhals._arrow.dataframe import ArrowDataFrame
+
+        return ArrowDataFrame(
+            native_dataframe=self._native_frame.arrow(),
+            backend_version=parse_version(pa.__version__),
+            version=self._version,
+        )
```

Comment:
My understanding was that duckdb is dependency free.
Should we jump to the discussion in #1479 before deciding how to _collect_ for duckdb?

Reply:
I think even in that one the likely default for duckdb would still be pyarrow though, right? i've added a try-except anyway to show a less surprising error message

Reply:
I will try to make a draft/RFC tomorrow to follow up on my comment in the thread ;)